### PR TITLE
feat(rust,python): optionally treat missing UTF8 values as the empty string at CSV parse-time

### LINKED ISSUE #5984: Add an option to exclude default null values when parsing CSV files
### Problem description

`pandas.read_csv` has an option `keep_default_na` which allows users to exclude default null values (by `keep_default_na=False`). I do think this is an important option that should added into Polars. A related issue is https://github.com/pola-rs/polars/issues/2769. Actually, sometimes, people do want to parse empty fields as empty strings instead of null values. However, there's no way to exclude null values in Polars currently.

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #5984.

# Explanation

When reading CSV data there is currently no easy/performant way to distinguish between `null` values that match an explicit list, and `null` values that come from _missing_ values; once loaded, they are all just `null`. Consequently it makes sense to be able to distinguish between them at parse-time (as long as there is no associated overhead).

Pandas supports this through a fairly [convoluted](https://github.com/pola-rs/polars/issues/5984#issuecomment-1368870853) interaction between `na_values` and `keep_na_values`, which is definitely not something we'd want to copy.

# Implementation

Aside from utf8 values, there don't really seem to be any other types that would have a reasonable non-null missing value, so I've exposed it on the CSV-reading interface as `missing_utf8_is_empty_string`, which makes it clear what it does and what it applies to (lower down this param name becomes the slightly more generic `missing_is_null`, which is always True _unless_ the new param flips it for utf8).

Essentially zero overhead; the boolean is passed-through the code until it can be used to flip the validity bitmap from False to True for the given missing value; no buffer-related shenanigans, etc. The only extra operation is a single `field.is_empty()` check that is triggered _iif_ the new flag is enabled and we've _already_ identified a potential null.

Added several new unit tests that try to stress the potential edge-cases...

# Example

```python
from textwrap import dedent
from io import StringIO
import polars as pl

csv = StringIO(dedent(
    r"""
    a,b,c,d,e,f,g
    na,,,,\N,,
    a,\N,c,,,,g
    ,,,,,,
    ,,,na,,,
    """
))
```
**Usual behaviour**: missing values and explicit null values _all_ become `null`:
```python
pl.read_csv( 
    csv,
    null_values = ["na",r"\N"],
)
# ┌──────┬──────┬──────┬──────┬──────┬──────┬──────┐
# │ a    ┆ b    ┆ c    ┆ d    ┆ e    ┆ f    ┆ g    │
# ╞══════╪══════╪══════╪══════╪══════╪══════╪══════╡
# │ null ┆ null ┆ null ┆ null ┆ null ┆ null ┆ null │
# │ a    ┆ null ┆ c    ┆ null ┆ null ┆ null ┆ g    │
# │ null ┆ null ┆ null ┆ null ┆ null ┆ null ┆ null │
# │ null ┆ null ┆ null ┆ null ┆ null ┆ null ┆ null │
# └──────┴──────┴──────┴──────┴──────┴──────┴──────┘
```
**Optional new behaviour**: missing values can be read as the empty string, with _only_ the explicitly-indicated values becoming `null`:
```python
pl.read_csv(
    csv,
    null_values = ["na",r"\N"],
    missing_utf8_is_empty_string = True,
)
# ┌──────┬──────┬─────┬──────┬──────┬──────┬─────┐
# │ a    ┆ b    ┆ c   ┆ d    ┆ e    ┆ f    ┆ g   │
# ╞══════╪══════╪═════╪══════╪══════╪══════╪═════╡
# │ null ┆      ┆     ┆      ┆ null ┆      ┆     │
# │ a    ┆ null ┆ c   ┆      ┆      ┆      ┆ g   │
# │      ┆      ┆     ┆      ┆      ┆      ┆     │
# │      ┆      ┆     ┆ null ┆      ┆      ┆     │
# └──────┴──────┴─────┴──────┴──────┴──────┴─────┘
```

### CODE REVIEW

File: polars/polars-io/src/csv/buffer.rs [hunk 1]
```diff
@@ -161,11 +168,12 @@ impl ParsedBuffer for Utf8Field {
         bytes: &[u8],
         ignore_errors: bool,
         needs_escaping: bool,
+        missing_is_null: bool,
     ) -> PolarsResult<()> {
         if bytes.is_empty() {
             // append null
             self.offsets.push(self.data.len() as i64);
-            self.validity.push(false);
+            self.validity.push(!missing_is_null);
```
Comment (ritchie46) [inline polars/polars-io/src/csv/buffer.rs:176]: That's a smart one. Makes it branchless indeed. :)

**Review by ritchie46 — COMMENTED** [review]
Looks good @alexander-beedie. I indeed expect this to not have a performance penalty.. Well done! :+1:

File: polars/polars-io/src/csv/buffer.rs (hunk 1 above)
Comment (alexander-beedie) [reply polars/polars-io/src/csv/buffer.rs:176]: At the moment my approach to Rust is _"run the code in my head as best I can, and stare at it until it bends to my will"_ 🤣 (maybe I'll ask on the Discord if anyone else has set up a Clion build env; would be nice to be able to step through and get a better sense of it).
